test/unit/storage/MonitoringStore.test.ts

```diff
@@ -85,9 +99,10 @@ describe('A MonitoringStore', (): void => {
     const result = store.deleteResource({ path: 'http://example.org/foo/bar' });
     expect(changedCallback).toHaveBeenCalledTimes(0);
     await result;
-    expect(changedCallback).toHaveBeenCalledTimes(2);
-    expect(changedCallback).toHaveBeenCalledWith({ path: 'http://example.org/modified/1' });
-    expect(changedCallback).toHaveBeenCalledWith({ path: 'http://example.org/modified/2' });
+    expect(changedCallback).toHaveBeenCalledTimes(3);
+    expect(changedCallback).toHaveBeenCalledWith({ path: 'http://example.org/modified/1' }, AS.Create);
+    expect(changedCallback).toHaveBeenCalledWith({ path: 'http://example.org/modified/2' }, AS.Update);
+    expect(changedCallback).toHaveBeenCalledWith({ path: 'http://example.org/modified/3' }, AS.Delete);
```

Comment:
The find the test title a bit misleading since I wouldn't expect *all* events to be fired by calling `deleteResource`. It may happen, but that's quite unlikely.

Perhaps it's better to change the test title to `it can fire all modified change events after deleteResource.` to better indicate that it's a possibility and not guaranteed.

Reply:
In this test we would expect all these to be definetly fired as this is how the child store (`source`) is mocked.

I think a better solution might be to change the mocked child store to return more appropriate data for each function. 

What do you think ?

Reply:
Indeed, the test is right accordingly to the mocked source.

However, I think my main issue with the current test title, is that it gives the wrong impression that a `MonitoringStore` is expected to always fire all events. I believe the test(s) could be improved in phrasing to serve as a nice specification for the developer how the component is generally expected to behave.

You might find this a bit excessive, but I do think by adding a nested `describe('event emission')` block would better convey the expected behavior of various situations to the developer.

```
describe('event emission', (): void => {
    const storeResult = jest.fn();
    const changedCb = jest.fn();
    const createdCb = jest.fn();
    const deletedCb = jest.fn();
    const updatedCb = jest.fn();
    let monitoringStore: MonitoringStore;

    beforeAll((): void => {
      const source2 = {
        getRepresentation: jest.fn(async(): Promise<any> => ({ success: true })),
        addResource: jest.fn(storeResult),
        setRepresentation: jest.fn(storeResult),
        deleteResource: jest.fn(storeResult),
        modifyResource: jest.fn(storeResult),
        hasResource: jest.fn(async(): Promise<any> => undefined),
      };
      monitoringStore = new MonitoringStore(source2);
      monitoringStore.on('changed', changedCb);
      monitoringStore.on('created', createdCb);
      monitoringStore.on('deleted', deletedCb);
      monitoringStore.on('updated', updatedCb);
    });

    beforeEach((): void => {
      jest.clearAllMocks();
    });

    it('fires no events if no resources was affected in the store.', async(): Promise<void> => {
      storeResult.mockImplementationOnce((): any => ({}));

      await monitoringStore.addResource({ path: 'http://example.org/foo/bar' }, {} as Patch);

      expect(changedCb).toHaveBeenCalledTimes(0);
      expect(createdCb).toHaveBeenCalledTimes(0);
      expect(deletedCb).toHaveBeenCalledTimes(0);
      expect(updatedCb).toHaveBeenCalledTimes(0);
    });

    it('fires fires "changed" and "created" events if a resource was created in the store.', async(): Promise<void> => {
      storeResult.mockImplementationOnce((): any => ({
        'http://example.org/foo/bar': new RepresentationMetadata({ path: 'http://example.org/foo/bar' })
          .add(SOLID_AS.terms.Activity, AS.Create),
      }));

      await monitoringStore.addResource({ path: 'http://example.org/foo/bar' }, {} as Patch);

      expect(changedCb).toHaveBeenCalledTimes(1);
      expect(changedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar' }, AS.Create);
      expect(createdCb).toHaveBeenCalledTimes(1);
      expect(createdCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar' });
      expect(deletedCb).toHaveBeenCalledTimes(0);
      expect(updatedCb).toHaveBeenCalledTimes(0);
    });

    it('fires fires "changed" and "updated" events if a resource was updated in the store.', async(): Promise<void> => {
      storeResult.mockImplementationOnce((): any => ({
        'http://example.org/foo/bar': new RepresentationMetadata({ path: 'http://example.org/foo/bar' })
          .add(SOLID_AS.terms.Activity, AS.Update),
      }));

      await monitoringStore.modifyResource({ path: 'http://example.org/foo/bar' }, {} as Patch);

      expect(changedCb).toHaveBeenCalledTimes(1);
      expect(changedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar' }, AS.Update);
      expect(createdCb).toHaveBeenCalledTimes(0);
      expect(deletedCb).toHaveBeenCalledTimes(0);
      expect(updatedCb).toHaveBeenCalledTimes(1);
      expect(updatedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar' });
    });

    it('fires fires "changed" and "deleted" events if a resource was delete in the store.', async(): Promise<void> => {
      storeResult.mockImplementationOnce((): any => ({
        'http://example.org/foo/bar': new RepresentationMetadata({ path: 'http://example.org/foo/bar' })
          .add(SOLID_AS.terms.Activity, AS.Delete),
      }));

      await monitoringStore.modifyResource({ path: 'http://example.org/foo/bar' }, {} as Patch);

      expect(changedCb).toHaveBeenCalledTimes(1);
      expect(changedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar' }, AS.Delete);
      expect(createdCb).toHaveBeenCalledTimes(0);
      expect(deletedCb).toHaveBeenCalledTimes(1);
      expect(deletedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar' });
      expect(updatedCb).toHaveBeenCalledTimes(0);
    });

    it('can fire combination of events if many resources was affected in the store.', async(): Promise<void> => {
      storeResult.mockImplementationOnce((): any => ({
        'http://example.org/foo/': new RepresentationMetadata({ path: 'http://example.org/foo/' })
          .add(SOLID_AS.terms.Activity, AS.Update),
        'http://example.org/foo/bar/': new RepresentationMetadata({ path: 'http://example.org/foo/bar/' })
          .add(SOLID_AS.terms.Activity, AS.Create),
        'http://example.org/foo/bar/baz': new RepresentationMetadata({ path: 'http://example.org/foo/bar/baz' })
          .add(SOLID_AS.terms.Activity, AS.Create),
      }));

      await monitoringStore.modifyResource({ path: 'http://example.org/foo/bar/baz' }, {} as Patch);

      expect(changedCb).toHaveBeenCalledTimes(3);
      expect(changedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/' }, AS.Update);
      expect(changedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar/' }, AS.Create);
      expect(changedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar/baz' }, AS.Create);
      expect(createdCb).toHaveBeenCalledTimes(2);
      expect(createdCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar/' });
      expect(createdCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/bar/baz' });
      expect(deletedCb).toHaveBeenCalledTimes(0);
      expect(updatedCb).toHaveBeenCalledTimes(1);
      expect(updatedCb).toHaveBeenCalledWith({ path: 'http://example.org/foo/' });
    });
  });
```

What do you think about this suggestion?

Reply:
I indeed find this a bit excessive :) But it could indeed be a great help to someone learning about the MonitoringStore.
I will start making these changes today, thanks for example codeblock btw!

Reply:
> Did you mean to have the old fires ___ events after ___ tests removed ?

Ah right, yes the tests from my previous example would essentially be a replacement for those tests.

> The suggested describe block seems to just be tests for the private function emitChanged. It's weird to me that we would create a whole describe block for a private function.

I suppose my reasoning is that the private function `emitChanged` is part of and affects the result/output when calling addResource/modifyResource/deleteResource. So it it's behavior should be tested.

But yeah, my example is quite excessive/too much. I think your suggested compromise is suitable. 👍